src/ray/core_worker/task_event_buffer.cc

```diff
@@ -0,0 +1,193 @@
+  if (!status.ok()) {
+    RAY_LOG(ERROR) << "Failed to connect to GCS, TaskEventBuffer will stop now. [status="
+                   << status.ToString() << "].";
+
+    // Early clean up resources.
+    gcs_client_.reset();
+    Stop();
+    return status;
+  }
+
+  if (!auto_flush) {
+    return Status::OK();
+  }
+
+  RAY_LOG(INFO) << "Reporting task events to GCS every " << report_interval_ms << "ms.";
+  periodical_runner_.RunFnPeriodically([this] { FlushEvents(/* forced */ false); },
+                                       report_interval_ms,
+                                       "CoreWorker.deadline_timer.flush_task_events");
+  return Status::OK();
+}
+
+void TaskEventBufferImpl::Stop() {
+  RAY_LOG(INFO) << "Shutting down TaskEventBuffer.";
+
+  // Shutting down the io service to exit the io_thread. This should prevent
+  // any other callbacks to be run on the io thread.
+  io_service_.stop();
+  if (io_thread_.joinable()) {
+    RAY_LOG(DEBUG) << "Joining io thread from TaskEventBuffer";
+    io_thread_.join();
+  }
+
+  {
+    absl::MutexLock lock(&mutex_);
+    // It's now safe to disconnect the GCS client since it will not be used by any
+    // callbacks.
+    if (gcs_client_) {
+      // Stop GCS client
+      gcs_client_->Disconnect();
+    }
+
+    gcs_client_.reset();
+  }
+}
+
+void TaskEventBufferImpl::AddTaskEvents(rpc::TaskEvents task_events) {
+  absl::MutexLock lock(&mutex_);
+
+  auto limit = RayConfig::instance().task_events_max_num_task_events_in_buffer();
+  if (limit > 0 && buffer_.size() >= static_cast<size_t>(limit)) {
+    // Too many task events, start overriding older ones.
+    buffer_[next_idx_to_overwrite_] = std::move(task_events);
+    next_idx_to_overwrite_ = (next_idx_to_overwrite_ + 1) % limit;
+    num_task_events_dropped_++;
+    return;
+  }
+  buffer_.push_back(std::move(task_events));
+}
+
+void TaskEventBufferImpl::FlushEvents(bool forced) {
+  std::vector<rpc::TaskEvents> task_events;
+  size_t num_task_events_dropped = 0;
+  {
+    absl::MutexLock lock(&mutex_);
+
+    RAY_LOG_EVERY_MS(INFO, 15000)
+        << "Pushed task state events to GCS. [total_bytes="
+        << (1.0 * total_events_bytes_) / 1024 / 1024
+        << "MiB][total_count=" << total_num_events_
+        << "][total_task_events_dropped=" << num_task_events_dropped_
+        << "][cur_buffer_size=" << buffer_.size() << "].";
+
+    // Skip if GCS hasn't finished processing the previous message.
+    if (grpc_in_progress_ && !forced) {
+      RAY_LOG_EVERY_N_OR_DEBUG(WARNING, 100)
+          << "GCS hasn't replied to the previous flush events call (likely overloaded). "
+             "Skipping reporting task state events and retry later."
+          << "[cur_buffer_size=" << buffer_.size() << "].";
+      return;
+    }
+
+    if (buffer_.size() == 0) {
+      return;
+    }
+
+    task_events.reserve(
+        RayConfig::instance().task_events_max_num_task_events_in_buffer());
+    buffer_.swap(task_events);
```

Comment:
Actually should we just not drop the events until the RPC succeeds? 
1. No additional complexity. There's no need to implement retry or anything. It is already handled. 
2. More resilient than losing the data upon RPC failures 

Is there any caveat in mind?

Reply:
One drawback is that we will then copy the data when we send to GCS. But given it's now in another thread, it's probably fine.

Reply:
Hmm maybe we should move it to another buffer and move it back if RPC fails. I think it might be okay not to handle it then if you think it is complicated. The only time I saw RPC failure in Ray is when GCS is down. Maybe we can do it as TODO later. 

If we don't handle RPC failure, you may need to include # of events to the "dropped events" when the RPC fail in this case though.

Reply:
Right - I think a failure in rpc usually indicate soemthing more severe than data loss here. (The dropped event count will also not be propogated to gcs based on current implementation) 

Yeah, I think what u mentioned should be part of the GCS client's logic for all rpc calls, it should have a retry logic.

Reply:
maybe we can just add 

if (!status.ok())

{
    data_loss += len(events)
}